snuba/clickhouse/astquery.py

```diff
@@ -0,0 +1,163 @@
+        # have to be removed from the Snuba Query
+        self.__sample = query.get_sample()
+        self.__hastotals = query.has_totals()
+
+        self.__settings = settings
+        self.__formatted_query: Optional[str] = None
+
+    def format_sql(self) -> str:
+        if self.__formatted_query:
+            return self.__formatted_query
+
+        parsing_context = ParsingContext()
+
+        selected_cols = [
+            e.accept(ClickhouseExpressionFormatter(parsing_context))
+            for e in self.__selected_columns
+        ]
+        formatted_columns = ", ".join(selected_cols)
+        select_clause = f"SELECT {formatted_columns}"
+
+        # TODO: The visitor approach will be used for the FROM clause as well.
+        from_clause = f"FROM {self.__data_source.format_from()}"
+
+        if self.__final:
+            from_clause = f"{from_clause} FINAL"
+
+        # TODO: Sampling rate will become one step of Clickhouse query processing
+        if not self.__data_source.supports_sample():
+            sample_rate = None
+        else:
+            if self.__sample:
+                sample_rate = self.__sample
+            elif self.__settings.get_turbo():
+                sample_rate = settings.TURBO_SAMPLE_RATE
+            else:
+                sample_rate = None
+        if sample_rate:
+            from_clause = f"{from_clause} SAMPLE {sample_rate}"
+
+        join_clause = ""
+        if self.__arrayjoin:
+            formatted_array_join = self.__arrayjoin.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            join_clause = f"ARRAY JOIN {formatted_array_join}"
+
+        where_clause = ""
+        if self.__condition:
+            formatted_condition = self.__condition.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            where_clause = f"WHERE {formatted_condition}"
+
+        # TODO: Pre where processing will become a step in Clickhouse Query processing
+        # instead of being pulled from the Snuba Query
+        prewhere_clause = ""
+        if self.__prewhere:
+            formatted_prewhere = self.__prewhere.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            prewhere_clause = f"PREWHERE {formatted_prewhere}"
+
+        group_clause = ""
+        if self.__groupby:
+            # reformat to use aliases generated during the select clause formatting.
+            groupby_expressions = [
+                e.accept(ClickhouseExpressionFormatter(parsing_context))
+                for e in self.__groupby
+            ]
+            formatted_groupby = ", ".join(groupby_expressions)
+            group_clause = f"GROUP BY ({formatted_groupby})"
+            if self.__hastotals:
+                group_clause = f"{group_clause} WITH TOTALS"
+
+        having_clause = ""
+        if self.__having:
+            assert self.__groupby, "found HAVING clause with no GROUP BY"
```

Comment:
Should this be validated in the constructor, rather than waiting until formatting to raise?

Reply:
Sure, just copied over from the old query processing.